node/network/statement-distribution/src/lib.rs

```diff
@@ -831,94 +847,132 @@ async fn handle_network_update(
-	let mut peers: HashMap<PeerId, PeerData> = HashMap::new();
-	let mut our_view = View::default();
-	let mut active_heads: HashMap<Hash, ActiveHeadData> = HashMap::new();
-	let mut statement_listeners: Vec<mpsc::Sender<SignedFullStatement>> = Vec::new();
-
-	loop {
-		let message = ctx.recv().await?;
-		match message {
-			FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate { activated, .. })) => {
-				for relay_parent in activated {
-					let (validators, session_index) = {
-						let (val_tx, val_rx) = oneshot::channel();
-						let (session_tx, session_rx) = oneshot::channel();
-
-						let val_message = AllMessages::RuntimeApi(
-							RuntimeApiMessage::Request(
-								relay_parent,
-								RuntimeApiRequest::Validators(val_tx),
-							),
-						);
-						let session_message = AllMessages::RuntimeApi(
-							RuntimeApiMessage::Request(
-								relay_parent,
-								RuntimeApiRequest::SessionIndexForChild(session_tx),
-							),
-						);
-
-						ctx.send_messages(
-							std::iter::once(val_message).chain(std::iter::once(session_message))
-						).await?;
-
-						match (val_rx.await?, session_rx.await?) {
-							(Ok(v), Ok(s)) => (v, s),
-							(Err(e), _) | (_, Err(e)) => {
-								log::warn!(
-									target: "statement_distribution",
-									"Failed to fetch runtime API data for active leaf: {:?}",
-									e,
-								);
-
-								// Lacking this bookkeeping might make us behave funny, although
-								// not in any slashable way. But we shouldn't take down the node
-								// on what are likely spurious runtime API errors.
-								continue;
+impl StatementDistribution {
+	async fn run(
+		self,
+		mut ctx: impl SubsystemContext<Message = StatementDistributionMessage>,
+	) -> SubsystemResult<()> {
+		let mut peers: HashMap<PeerId, PeerData> = HashMap::new();
+		let mut our_view = View::default();
+		let mut active_heads: HashMap<Hash, ActiveHeadData> = HashMap::new();
+		let mut statement_listeners: Vec<mpsc::Sender<SignedFullStatement>> = Vec::new();
+		let metrics = self.metrics;
+
+		loop {
+			let message = ctx.recv().await?;
+			match message {
+				FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate { activated, .. })) => {
+					for relay_parent in activated {
+						let (validators, session_index) = {
+							let (val_tx, val_rx) = oneshot::channel();
+							let (session_tx, session_rx) = oneshot::channel();
+
+							let val_message = AllMessages::RuntimeApi(
+								RuntimeApiMessage::Request(
+									relay_parent,
+									RuntimeApiRequest::Validators(val_tx),
+								),
+							);
+							let session_message = AllMessages::RuntimeApi(
+								RuntimeApiMessage::Request(
+									relay_parent,
+									RuntimeApiRequest::SessionIndexForChild(session_tx),
+								),
+							);
+
+							ctx.send_messages(
+								std::iter::once(val_message).chain(std::iter::once(session_message))
+							).await?;
+
+							match (val_rx.await?, session_rx.await?) {
```

Comment:
Looks a bit like a candidate for `join!` macro if available.

Reply:
I haven't touched that code (it was only indented one level right as it became a method), but sure.

Reply:
Because of `?` after await, it's not that simple of a conversion to `join!` unless you know how to implement this easily.
If not, this is outside of the PR's scope :)

Reply:
this is what I came up with
```patch
diff --git a/node/network/statement-distribution/src/lib.rs b/node/network/statement-distribution/src/lib.rs
index 4dac16bc..db6e8587 100644
--- a/node/network/statement-distribution/src/lib.rs
+++ b/node/network/statement-distribution/src/lib.rs
@@ -884,7 +884,8 @@ impl StatementDistribution {
                                                                std::iter::once(val_message).chain(std::iter::once(session_message))
                                                        ).await?;
 
-                                                       match (val_rx.await?, session_rx.await?) {
+                                                       let (a, b) = futures::future::join(val_rx, session_rx).await;
+                                                       match (a?, b?) {
                                                                (Ok(v), Ok(s)) => (v, s),
                                                                (Err(e), _) | (_, Err(e)) => {
                                                                        log::warn!(

```
but tbh I don't see any benefit over the loss of readability

Reply:
You are right, there is no gain really.